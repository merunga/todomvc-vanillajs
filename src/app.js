let todos = [
  {
    id: Date.now(),
    label: 'Un todo desde el store',
  },
];

export const renderApp = () => {
  // Creamos el elemento padre
  const appElem = document.createElement('section');
  appElem.setAttribute('class', 'todoapp');

  // Creamos cada uno de los elementos hijos

  // HEADER
  const headerElem = document.createElement('header');
  headerElem.setAttribute('class', 'header');
  headerElem.innerHTML = `
    <h1>todos</h1>
    <input class="new-todo" placeholder="What needs to be done?" autofocus="">
  `;

  // Obtenemos el elemento input
  const inputElem = headerElem.querySelector('input');
  // Le asignamos una acccion al evento de 'tecla presionada'
  inputElem.addEventListener('keypress', (evt) => {
    // Es un ENTER ?
    if (evt.keyCode === 13) {
      const label = evt.target.value; // Obtenemos el valor del input
      if (label.trim()) { // si el input no esta vacio
        // Insertamos el nuevo todo
        const todo = {
          id: Date.now(), // creamos un id
          label,
        };
        todos = [
          ...todos,
          todo,
        ];

        renderApp(); // Volvemos a pintar la vista para que se reflejen los cambios
        document.querySelector('.new-todo').focus(); // Y volvemos a hacer foco en el input
      }
    }
  });

  // LISTA
  const todosElem = document.createElement('section');
  todosElem.setAttribute('class', 'main');

  const ulElem = document.createElement('div');
  ulElem.setAttribute('class', 'todo-list');

  todosElem.appendChild(ulElem);

  todos.forEach((todo) => {
    const todoElem = document.createElement('li');
    todoElem.setAttribute('data-id', todo.id);
    todoElem.innerHTML = `
      <div class="view">
        <label>${todo.label}</label>
        <button class="destroy"></button>
      </div>
    `;

    // obtenemos el boton de eliminar
    const buttonElem = todoElem.querySelector('button');
    // le asociamos una accion al evento 'click'
    buttonElem.addEventListener('click', (evt) => {
      // Obtenemos el elemento padre del boton, que engloba al todo
      // https://developer.mozilla.org/es/docs/Web/API/Element/closest
      const liElem = evt.target.closest('li');
      const id = parseInt(liElem.dataset.id);
      todos = todos.filter((t) => t.id !== id);

      renderApp(); // Volvemos a pintar la vista para que se reflejen los cambios
    });
    ulElem.appendChild(todoElem);
  });

  // FOOTER
  const footerElem = document.createElement('footer');
  footerElem.setAttribute('class', 'footer');
  footerElem.innerHTML = `<span class="todo-count"><strong>${todos.length}</strong> todo</span>`;

  // Los insertamos en el padre
  appElem.appendChild(headerElem);
  appElem.appendChild(todosElem);
  appElem.appendChild(footerElem);

  const root = document.getElementById('root');
  root.innerHTML = '';
  // lo insertamos en el elemento #root
  root.appendChild(appElem);
};
