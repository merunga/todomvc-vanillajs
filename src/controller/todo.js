import { uuid } from './utils.js';

/**
 * Inserta un nuevo todo al final del original
 * @param {*} todos Array en donde insertar el nuevo todo
 * @param {*} label Label del nuevo todo
 */
export const todoInsertar = (todos, label) => {
  const todo = {
    id: uuid(),
    label,
  };
  return [
    ...todos,
    todo,
  ];
};

/**
 * Elimina un todo del array basado en el id
 * @param {*} todos Array de donde quitar el todo
 * @param {*} id Id del todo a eliminar
 */
export const todoEliminar = (todos, id) => (
  todos.filter((t) => t.id !== id)
);
