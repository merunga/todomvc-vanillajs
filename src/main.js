import { renderApp } from './view/index.js';
import { createStore } from './controller/store.js';

createStore({
  todos: [
    {
      id: 1,
      label: 'Un todo desde el store',
    },
  ],
});

window.addEventListener('load', () => {
  renderApp();
});
