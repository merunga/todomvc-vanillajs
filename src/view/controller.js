import store from '../controller/store.js';
import App from './App.js';

/**
 * Crea el elemento principal y lo inserta en el DOM dentro del elemeno #root
 */
export const renderApp = () => {
  // Obtenemos los todos desde el store
  const todos = store.get('todos');

  const app = App(todos);

  const root = document.getElementById('root');
  root.innerHTML = '';
  // lo insertamos en el elemento #root
  root.appendChild(app);
};
