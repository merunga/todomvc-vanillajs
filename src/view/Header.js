import store from '../controller/store.js';
import { todoInsertar } from '../controller/todo.js';
import { renderApp } from './index.js';

export const insertarOnEnter = (evt) => {
  // Es un ENTER ?
  if (evt.keyCode === 13) {
    const label = evt.target.value; // Obtenemos el valor del input
    if (label.trim()) { // si el input no esta vacio
      const todos = store.get('todos'); // Obtenemos los todos actuales
      const newTodos = todoInsertar(todos, label); // Insertamos el nuevo todo
      store.set('todos', newTodos); // Lo guardamos en el store

      renderApp(); // Volvemos a pintar la vista para que se reflejen los cambios
      document.querySelector('.new-todo').focus(); // Y volvemos a hacer foco en el input
    }
  }
};

/**
 * Crea el elemento del DOM de la cabecera
 */
export default () => {
  const header = document.createElement('header');
  header.setAttribute('class', 'header');
  header.innerHTML = `
    <h1>todos</h1>
    <input class="new-todo" placeholder="What needs to be done?" autofocus="">
  `;

  // Obtenemos el elemento input
  const inputElem = header.querySelector('input');
  // Le asignamos una acccion al evento de 'tecla presionada'
  inputElem.addEventListener('keypress', insertarOnEnter);

  return header;
};
