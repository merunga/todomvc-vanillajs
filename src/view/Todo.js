import store from '../controller/store.js';
import { todoEliminar } from '../controller/todo.js';
import { renderApp } from './index.js';

export const eliminarOnClick = (evt) => {
  // Obtenemos el elemento padre del boton, que engloba al todo
  // https://developer.mozilla.org/es/docs/Web/API/Element/closest
  const liElem = evt.target.closest('li');
  const id = parseInt(liElem.dataset.id);

  const todos = store.get('todos'); // Obtenemos los todos actuales
  const newTodos = todoEliminar(todos, id); // Eliminamos el todo
  store.set('todos', newTodos); // Guardamos el array actualizando en el store

  renderApp(); // Volvemos a pintar la vista para que se reflejen los cambios
};

/**
 * Crea el elemento del DOM para un `todo`
 *
 * @param todo Objecto con las propiedades `id` y `label`
 */
export default (todo) => {
  const liElem = document.createElement('li');
  liElem.setAttribute('data-id', todo.id);
  liElem.innerHTML = `
    <div class="view">
      <label>${todo.label}</label>
      <button class="destroy"></button>
    </div>
  `;

  // obtenemos el boton de eliminar
  const buttonElem = liElem.querySelector('button');
  // le asociamos una accion al evento 'click'
  buttonElem.addEventListener('click', eliminarOnClick);

  return liElem;
};
