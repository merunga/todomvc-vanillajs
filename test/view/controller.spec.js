import { createStore } from '../../src/controller/store';
import { renderApp } from '../../src/view';

describe('View controller', () => {
  beforeEach(() => {
    document.body.innerHTML = '<div id="root"></div>';
  });
  it('renderApp inserta app dentro del elemento #root', () => {
    createStore({
      todos: [
        {
          id: 'id',
          label: 'Un todo',
        },
      ],
    });
    renderApp();
    const rootElem = document.getElementById('root');
    // Mas info de snapshots en https://jestjs.io/docs/es-ES/snapshot-testing
    expect(rootElem.innerHTML).toMatchSnapshot();
  });
});
