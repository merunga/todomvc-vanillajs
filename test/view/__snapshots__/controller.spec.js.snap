// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View controller renderApp inserta app dentro del elemento #root 1`] = `
"<section class=\\"todoapp\\"><header class=\\"header\\">
    <h1>todos</h1>
    <input class=\\"new-todo\\" placeholder=\\"What needs to be done?\\" autofocus=\\"\\">
  </header><section class=\\"main\\"><div class=\\"todo-list\\"><li data-id=\\"id\\">
    <div class=\\"view\\">
      <label>Un todo</label>
      <button class=\\"destroy\\"></button>
    </div>
  </li></div></section><footer class=\\"footer\\"><span class=\\"todo-count\\"><strong>1</strong> todo</span></footer></section>"
`;
